### DISCUSSION TITLE
problem with laser control

### DISCUSSION THREAD

**Discussion by patrik156** [arkypita/LaserGRBL#877] [Q&A]
please can someone help me with my laser. i have problem with laser control, when i set 
M3 S0 i have max power
M3 S1000 is low
how i can change it
i use grbl1.1 , mana SE 3.2 LASER 5,5 w and TTL to PWM modul

**Comment by patrik156** [comment]
![84702803_565311637389075_2324459003792326656_n](https://user-images.githubusercontent.com/60711808/74419249-e6cd1300-4e49-11ea-84e3-594a9a3b3c9b.jpg)

**Comment by patrik156** [comment]
![84680079_632875403953712_2440349071253176320_n](https://user-images.githubusercontent.com/60711808/74419315-fea49700-4e49-11ea-88e1-18ebfbf8fc61.jpg)

**Comment by arkypita** [comment]
Hi @patrik156 
In the picture of your laser driver I read PWM=0 : ON, PWM=1: OFF.

This is the opposite of the standard, this is the opposit of the logical signal your manaSE board generate.

You can reverse signal using a transistor (any NPN transistor is ok) as digital signal inverter between manaSE board and laser driver.

![image](https://user-images.githubusercontent.com/8782035/74419882-d9fcef00-4e4a-11ea-85fd-4131a420a3cf.png)

Or if you want a software solution you should re-compile and flash grbl to your arduino control board. There is a [macro in grbl config.h](https://github.com/gnea/grbl/blob/eefe2bb95bb7b21ec2bb87e6ab6e20747e1626c4/grbl/config.h#L205) called `INVERT_SPINDLE_ENABLE_PIN `that (if uncommented) allow to invert logic inside the control board.

**Comment by patrik156** [comment]
https://github.com/gnea/grbl/releases/tag/v1.1h.20190825 
i download and uncommented INVERT_SPINDLE_ENABLE_PIN but now i cant  upload i have this message 
Sketch uses 29762 bytes (96%) of program storage space. Maximum is 30720 bytes.
Global variables use 1633 bytes (79%) of dynamic memory, leaving 415 bytes for local variables. Maximum is 2048 bytes.
Low memory available, stability problems may occur.
 before this i use grbl laser for upload

**Comment by arkypita** [comment]
it is perfectly normal

**Comment by arkypita** [comment]
https://github.com/gnea/grbl/wiki/Frequently-Asked-Questions#compiling-grbl

**Comment by patrik156** [comment]
i still have roblem , can you help me 
In file included from C:\Users\PC1\Documents\Arduino\libraries\grbl/grbl.h:43:0,

                 from C:\Users\PC1\Documents\Arduino\libraries\grbl\examples\grblUpload\grblUpload.ino:27:

C:\Users\PC1\Documents\Arduino\libraries\grbl/nuts_bolts.h:56:0: warning: "max" redefined

 #define max(a,b) (((a) > (b)) ? (a) : (b))

 

In file included from sketch\grblUpload.ino.cpp:1:0:

C:\Users\PC1\Desktop\arduino-1.8.11-windows\arduino-1.8.11\hardware\arduino\avr\cores\arduino/Arduino.h:93:0: note: this is the location of the previous definition

 #define max(a,b) ((a)>(b)?(a):(b))

 

In file included from C:\Users\PC1\Documents\Arduino\libraries\grbl/grbl.h:43:0,

                 from C:\Users\PC1\Documents\Arduino\libraries\grbl\examples\grblUpload\grblUpload.ino:27:

C:\Users\PC1\Documents\Arduino\libraries\grbl/nuts_bolts.h:57:0: warning: "min" redefined

 #define min(a,b) (((a) < (b)) ? (a) : (b))

 

In file included from sketch\grblUpload.ino.cpp:1:0:

C:\Users\PC1\Desktop\arduino-1.8.11-windows\arduino-1.8.11\hardware\arduino\avr\cores\arduino/Arduino.h:92:0: note: this is the location of the previous definition

 #define min(a,b) ((a)<(b)?(a):(b))

 

In file included from C:\Users\PC1\Documents\Arduino\libraries\grbl/grbl.h:43:0,

                 from C:\Users\PC1\Documents\Arduino\libraries\grbl\examples\grblUpload\grblUpload.ino:27:

C:\Users\PC1\Documents\Arduino\libraries\grbl/nuts_bolts.h:61:0: warning: "bit" redefined

 #define bit(n) (1 << n)

 

In file included from sketch\grblUpload.ino.cpp:1:0:

C:\Users\PC1\Desktop\arduino-1.8.11-windows\arduino-1.8.11\hardware\arduino\avr\cores\arduino/Arduino.h:124:0: note: this is the location of the previous definition

 #define bit(b) (1UL << (b))

 

Sketch uses 29762 bytes (96%) of program storage space. Maximum is 30720 bytes.
Global variables use 1633 bytes (79%) of dynamic memory, leaving 415 bytes for local variables. Maximum is 2048 bytes.
Low memory available, stability problems may occur.

**Comment by arkypita** [comment]
These question should be done on the github page of grbl project, not here (LaserGRBL project).
First, however, do a search on old closed issue in grbl.... it has already been requested!

https://github.com/gnea/grbl/issues?utf8=✓&q=warning+redefined

**Comment by patrik156** [comment]
yes i find it , but now have another problem , i uncomment INVERT_SPINDLE_ENABLE_PIN , but it doesnt help

**Comment by StuartB4** [comment]
> Or if you want a software solution you should re-compile and flash grbl to your arduino control board. There is a macro in grbl config.h called INVERT_SPINDLE_ENABLE_PIN that (if uncommented) allow to invert logic inside the control board.

Just had to do that myself. Bought a new 5 watt laser from wish.com but it has the TTL signal reversed on the built in driver board. So when it's powered and the PWM signal from the Arduino is 0 the laser comes on and stays on. I enabled the INVERT_SPINDLE_ENABLE_PIN in config.h and compiled and re-uploaded grbl to the Arduino and it works perfectly.

**Comment by StuartB4** [comment]
Patrik, did you compile and upload grbl to the arduino after making the invert spindle change?

**Comment by patrik156** [comment]
yes , but still have max power with M3 S0 

![sadfghjklôôn- ,](https://user-images.githubusercontent.com/60711808/74514587-2c064900-4f0d-11ea-82f2-d69fc755fe97.PNG)

**Comment by patrik156** [comment]
and this configuration is ok ?
![dgnxfgn](https://user-images.githubusercontent.com/60711808/74515071-278e6000-4f0e-11ea-8175-0bde864cfba1.PNG)

**Comment by patrik156** [comment]
![89461311_647495469151638_7874393323954241536_n](https://user-images.githubusercontent.com/60711808/76691128-c1e5de80-6647-11ea-9942-de6361d32201.jpg)
i try to connect NPN transistor , but i am not sure if it is correct . please can you check it please ?

**Comment by StuartB4** [comment]
The PWM signal should come out between Pin 3 on transistor and the end of the resistor and goto the PWM wire on the laser.
The other end of the resistor should have a voltage going in, 5 volts is enough.
![76691128-c1e5de80-6647-11ea-9942klkl-de6361d32201](https://user-images.githubusercontent.com/16980026/76691794-b219ca80-6646-11ea-8a4f-aa353f3f70d8.jpg)

**Comment by patrik156** [comment]
thank you ,now it works ,but i have one more question because  i think that my laser is weak . so i set M3 S1000 and  input to my laser is 5,5V . it is maximum ?

**Comment by StuartB4** [comment]
If you set $30 to 1000 in grbl settings then that will be the maximum laser power.  Any number in between that and the minimum setting on $31 (should be set to 0) will be lower power. So if you set $31 to 1000 in grbl and then set the laser power to 500 in Lasergrbl settings when you prepare an image,  then it will be 1/2 power, 250 will be 1/4 power and so on.
